proposals/XXXX-efficient-token-program.md

```diff
@@ -0,0 +1,97 @@
+
+Additionally, there are benefits to downstream programs:
+* Better composability since using the Token program instructions require less CUs.
+* Cheaper (in terms of CU) cross-program invocations.
+
+## New Terminology
+
+N/A.
+
+## Detailed Design
+
+`p-token` is a like-for-like efficient re-implementation of the Token program. It is `no_std` (no heap memory allocations are made in the program) and uses zero-copy access for instruction and account data. Since it follows the same instructions and accounts layout, it does not require any changes to client code – it works as a drop-in replacement.
+
+Apart from the original SPL Token instructions, this proposal adds two additional instructions to the program:
+
+1. [`withdraw_excess_lamports`](https://github.com/solana-program/token/blob/main/p-token/src/processor/withdraw_excess_lamports.rs) (instruction discriminator `38`): allow recovering "bricked" SOL from mint (e.g., USDC mint as `~323` SOL in excess) and multisig accounts. The logic of this instruction is the same as the current SPL Token-2022 instruction. 
+2. [`batch`](https://github.com/solana-program/token/blob/main/p-token/src/processor/batch.rs) (instruction discriminator `255`): enable efficient CPI interaction with the Token program. This is a new instruction that can execute a variable number on Token instructions in a single invocation of the Token program. Therefore, the CPI invoke units (currently `1000` CU) are only consumed once, instead of each CPI instruction – this significantly improves the CUs required to perform multiple Token instructions in a CPI context.
+
+Note that `withdraw_excess_lamports` discriminator matches the value used in SPL Token-2022, while `batch` has a driscriminator value that is not used in either SPL Token nor Token-2022.
+
+The program will be loaded into an account (`ptokNfvuU7terQ2r2452RzVXB3o4GT33yPWo1fUkkZ2`) prior to enabling the feature gate that triggers the replacement.
+
+When the feature gate `ptokSWRqZz5u2xdqMdstkMKpFurauUpVen7TZXgDpkQ` is enabled, the runtime needs to:
+
+* Replace `TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA` program with `ptokNfvuU7terQ2r2452RzVXB3o4GT33yPWo1fUkkZ2` using the Upgradable Loader `v4`.
+
+Another point for consideration is to whether to keep the current logs of the program or not. Logs are limited to show the name of the instruction being executed, e.g.:
+```
+Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA invoke [1]
+Program log: Instruction: Transfer
+Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA consumed 249 of 200000 compute units
+Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA success
+```
+Logging the instruction name consumes around `103` compute units, which in some cases reprensents almost the same amount required to execute the instruction. Note that errors will always be logged, since this is considered good use of CU.
```

Comment:
Failed transactions use up all CUs anyway, so we can be as wasteful as we want in failing :wink: